### ISSUE TITLE
[gofish] how to set BasicAuth for redfish provider

### ISSUE THREAD

**Issue by andy-vm** [bmc-toolbox/bmclib#328]
Hi, I encountered 401 error while using redfish provider, adding `BasicAuth: true` to config solved the issue, is there any way to pass the config down to gofish? thanks

```go
config := gofish.ClientConfig{
		Endpoint:   host,
		Username:   user,
		Password:   pass,
		Insecure:   true,
		HTTPClient: httpClient,
		BasicAuth:  true,
	}
```

```yml
Unauthorized
--- FAIL: TestBMC (0.57s)
    bmc_test.go:29: 
                Error Trace:    /home/github.com/-vm/pkg/bmc/bmc_test.go:29
                Error:          Not equal: 
                                expected: <nil>(<nil>)
                                actual  : *multierror.Error(*multierror.Error{Errors:[]error{provider: gofish: failed to retrieve some items: [{"link":"/redfish/v1/Systems/system","error":"401: Unauthorized"}], failed to get power state}, ErrorFormat:(multierror.ErrorFormatFunc)(nil)})
                Test:           TestBMC
                Messages:       BMC poweron https://10.2.7.5: failed: 2 errors occurred:
                                        * provider: gofish: failed to retrieve some items: [{"link":"/redfish/v1/Systems/system","error":"401: Unauthorized"}]
                                        * failed to get power state
```

**Comment by andy-vm** [comment]
to add more info, below is the log when I use `bmclib` w/o `BasicAuth` in gofish client config, is there any solution to solve the issue other than passing down the  `BasicAuth` to gofish, thanks.
 
```yml
GET /redfish/v1/ HTTP/1.1
Host: 10.22.7.5
User-Agent: gofish/1.0
Connection: close
Accept: application/json
Content-Type: application/json
Accept-Encoding: gzip


HTTP/1.1 200 OK
Connection: close
Content-Length: 1303
Cache-Control: no-Store,no-Cache
Content-Security-Policy: default-src 'none'; img-src 'self' data:; font-src 'self'; style-src 'self'; script-src 'self'; connect-src 'self' wss:; form-action 'none'; frame-ancestors 'none'; plugin-types 'none'; base-uri 'none'
Content-Type: application/json
Date: Wed, 03 May 2023 13:53:04 GMT
Pragma: no-cache
Strict-Transport-Security: max-age=31536000; includeSubdomains; preload
X-Content-Type-Options: nosniff
X-Frame-Options: DENY
X-Xss-Protection: 1; mode=block

{
  "@odata.id": "/redfish/v1",
  "@odata.type": "#ServiceRoot.v1_5_0.ServiceRoot",
  "AccountService": {
    "@odata.id": "/redfish/v1/AccountService"
  },
  "CertificateService": {
    "@odata.id": "/redfish/v1/CertificateService"
  },
  "Chassis": {
    "@odata.id": "/redfish/v1/Chassis"
  },
  "EventService": {
    "@odata.id": "/redfish/v1/EventService"
  },
  "Id": "RootService",
  "JsonSchemas": {
    "@odata.id": "/redfish/v1/JsonSchemas"
  },
  "Links": {
    "Sessions": {
      "@odata.id": "/redfish/v1/SessionService/Sessions"
    }
  },
  "Managers": {
    "@odata.id": "/redfish/v1/Managers"
  },
  "Name": "Root Service",
  "Oem": {
    "OpenBmc": {
      "@odata.type": "#OemServiceRoot.v1_0_0.ServiceRoot",
      "CupsService": {
        "@odata.id": "/redfish/v1/CupsService"
      }
    }
  },
  "RedfishVersion": "1.9.0",
  "Registries": {
    "@odata.id": "/redfish/v1/Registries"
  },
  "SessionService": {
    "@odata.id": "/redfish/v1/SessionService"
  },
  "Systems": {
    "@odata.id": "/redfish/v1/Systems"
  },
  "Tasks": {
    "@odata.id": "/redfish/v1/TaskService"
  },
  "TelemetryService": {
    "@odata.id": "/redfish/v1/TelemetryService"
  },
  "UUID": "87c34931-c0d1-54db-a5c9-6b97cd6d049f",
  "UpdateService": {
    "@odata.id": "/redfish/v1/UpdateService"
  }
}
POST /redfish/v1/SessionService/Sessions HTTP/1.1
Host: 10.22.7.5
User-Agent: gofish/1.0
Connection: close
Content-Length: 45
Accept: application/json
Content-Type: application/json
Accept-Encoding: gzip

{"UserName":"user","Password":"xxx"}
HTTP/1.1 201 Created
Connection: close
Content-Length: 268
Cache-Control: no-Store,no-Cache
Content-Security-Policy: default-src 'none'; img-src 'self' data:; font-src 'self'; style-src 'self'; script-src 'self'; connect-src 'self' wss:; form-action 'none'; frame-ancestors 'none'; plugin-types 'none'; base-uri 'none'
Content-Type: application/json
Date: Wed, 03 May 2023 13:53:04 GMT
Location: /redfish/v1/SessionService/Sessions/VF8KSqNWI1
Pragma: no-cache
Strict-Transport-Security: max-age=31536000; includeSubdomains; preload
X-Auth-Token: 7ix2wlLARkRa253ZDXlX
X-Content-Type-Options: nosniff
X-Frame-Options: DENY
X-Xss-Protection: 1; mode=block

{
  "@odata.id": "/redfish/v1/SessionService/Sessions/VF8KSqNWI1",
  "@odata.type": "#Session.v1_3_0.Session",
  "ClientOriginIPAddress": "10.226.76.8",
  "Description": "Manager User Session",
  "Id": "VF8KSqNWI1",
  "Name": "User Session",
  "UserName": "user"
}
GET /redfish/v1/Systems HTTP/1.1
Host: 10.22.7.5
User-Agent: gofish/1.0
Connection: close
Accept: application/json
Content-Type: application/json
Cookie: sessionKey=7ix2wlLARkRa253ZDXlX
X-Auth-Token: 7ix2wlLARkRa253ZDXlX
Accept-Encoding: gzip


HTTP/1.1 200 OK
Connection: close
Content-Length: 259
Cache-Control: no-Store,no-Cache
Content-Security-Policy: default-src 'none'; img-src 'self' data:; font-src 'self'; style-src 'self'; script-src 'self'; connect-src 'self' wss:; form-action 'none'; frame-ancestors 'none'; plugin-types 'none'; base-uri 'none'
Content-Type: application/json
Date: Wed, 03 May 2023 13:53:04 GMT
Pragma: no-cache
Strict-Transport-Security: max-age=31536000; includeSubdomains; preload
X-Content-Type-Options: nosniff
X-Frame-Options: DENY
X-Xss-Protection: 1; mode=block

{
  "@odata.id": "/redfish/v1/Systems",
  "@odata.type": "#ComputerSystemCollection.ComputerSystemCollection",
  "Members": [
    {
      "@odata.id": "/redfish/v1/Systems/system"
    }
  ],
  "Members@odata.count": 1,
  "Name": "Computer System Collection"
}
GET /redfish/v1/Systems/system HTTP/1.1
Host: 10.22.7.5
User-Agent: gofish/1.0
Connection: close
Accept: application/json
Content-Type: application/json
Cookie: sessionKey=7ix2wlLARkRa253ZDXlX
X-Auth-Token: 7ix2wlLARkRa253ZDXlX
Accept-Encoding: gzip


HTTP/1.1 401 Unauthorized
Connection: close
Content-Length: 12
Cache-Control: no-Store,no-Cache
Content-Security-Policy: default-src 'none'; img-src 'self' data:; font-src 'self'; style-src 'self'; script-src 'self'; connect-src 'self' wss:; form-action 'none'; frame-ancestors 'none'; plugin-types 'none'; base-uri 'none'
Date: Wed, 03 May 2023 13:53:04 GMT
Pragma: no-cache
Strict-Transport-Security: max-age=31536000; includeSubdomains; preload
X-Content-Type-Options: nosniff
X-Frame-Options: DENY
X-Xss-Protection: 1; mode=block

Unauthorized
DELETE /redfish/v1/SessionService/Sessions/VF8KSqNWI1 HTTP/1.1
Host: 10.22.7.5
User-Agent: gofish/1.0
Connection: close
Accept: application/json
Content-Type: application/json
Cookie: sessionKey=7ix2wlLARkRa253ZDXlX
X-Auth-Token: 7ix2wlLARkRa253ZDXlX
Accept-Encoding: gzip


HTTP/1.1 401 Unauthorized
Connection: close
Content-Length: 12
Cache-Control: no-Store,no-Cache
Content-Security-Policy: default-src 'none'; img-src 'self' data:; font-src 'self'; style-src 'self'; script-src 'self'; connect-src 'self' wss:; form-action 'none'; frame-ancestors 'none'; plugin-types 'none'; base-uri 'none'
Date: Wed, 03 May 2023 13:53:04 GMT
Pragma: no-cache
Strict-Transport-Security: max-age=31536000; includeSubdomains; preload
X-Content-Type-Options: nosniff
X-Frame-Options: DENY
X-Xss-Protection: 1; mode=block

Unauthorized
--- FAIL: TestBMC (0.57s)
    bmc_test.go:29: 
                Error Trace:    /home/github.com/vm/pkg/bmc/bmc_test.go:29
                Error:          Not equal: 
                                expected: <nil>(<nil>)
                                actual  : *multierror.Error(*multierror.Error{Errors:[]error{provider: gofish: failed to retrieve some items: [{"link":"/redfish/v1/Systems/system","error":"401: Unauthorized"}], failed to get power state}, ErrorFormat:(multierror.ErrorFormatFunc)(nil)})
                Test:           TestBMC
                Messages:       BMC poweron https://10.22.7.5: failed: 2 errors occurred:
                                        * provider: gofish: failed to retrieve some items: [{"link":"/redfish/v1/Systems/system","error":"401: Unauthorized"}]
                                        * failed to get power state
```

**Comment by joelrebel** [comment]
@andy-vm with #329 you can set basic auth

**Comment by andy-vm** [comment]
hi @joelrebel  thanks for the prompt update, I tried to fork this repo earlier and set basic auth to true, but it failed with error below, not sure if #329 will fix this issue as well?

```
*multierror.Error(*multierror.Error{Errors:[]error{provider: gofish: client not authenticated: not authenticated: not authenticated, failed to get power state}, ErrorFormat:(multierror.ErrorFormatFunc)(nil)})
```

**Comment by andy-vm** [comment]
I re-ran my test with #329 and confirmed that it fixed the issue, this ticket can be closed once #329 is merged, thx
